opacus/optimizers/optimizer.py

```diff
@@ -8,19 +8,60 @@
 
 
 def _generate_noise(
-    std: float, reference: torch.Tensor, generator=None
+    std: float,
+    reference: torch.Tensor,
+    generator=None,
+    secure_mode: bool = False,
 ) -> torch.Tensor:
-    if std > 0:
-        # TODO: handle device transfers: generator and reference tensor
-        # could be on different devices
+    """
+    Generates noise according to a Gaussian distribution with mean 0
+
+    Args:
+        std: Standard deviation of the noise
+        reference: The reference Tensor to get the appripriate shape and device
+            for generating the noise
+        generator: The PyTorch noise generator
+        secure_mode: boolean showing if "secure" noise need to be generate
+            (see the notes)
+
+    Notes:
+        If `secure_mode` is enabled, the generated noise is also secure
+        against the floating point representation attacks, such as the ones
+        in https://arxiv.org/abs/2107.10138. This is achieved through calling
+        the Gaussian noise function 2*n times, when n=2 (see section 5.1 in
+        https://arxiv.org/abs/2107.10138)
+    """
+    zeros = torch.zeros(reference.shape, device=reference.device)
+    if std == 0:
+        return zeros
+    # TODO: handle device transfers: generator and reference tensor
+    # could be on different devices
+    if secure_mode:
+        torch.normal(
+            mean=0,
+            std=std,
+            size=(1, 1),
```

Comment:
Dumb question: is it OK for the shape to be something different? I'm guessing yes, since you explicitly coded it this way, but wanted to confirm.

Reply:
This call to `torch.normal` doesn't seem to have any effect - the result it not stored or used anywhere, unless I've missed something

Reply:
Sorry for the confusion.

The purpose of this line of code is to generate one Gaussian sample and throw away (that's why it's not used).

According to the paper: "Readers are encouraged to *discard* the first result from Box-Muller before sampling in this way, to ensure no carry-over of uniform variates from a previous draw, which may have already been successfully attacked."

Reply:
I will make the comment more clear

Reply:
The fact that you were throwing it away was clear to me. What was not clear is if the shape can be any dumb thing. I am not familiar with how `torch.normal` is implemented and if using a different shape/size than the subsequent has any impact.

Reply:
Oh I see. The shape argument in `torch.normal` is to determine how many random numbers to generate. I am here generating one number by doing `(1,1)` @karthikprasad